rfcs/text/0016_buildkite.md

```diff
@@ -0,0 +1,923 @@
+
+![High-Level Design](../images/0016_agent_manager.png)
+
+The high-level design for the agent manager is pretty straightforward. There are three primary stages during execution:
+
+1. Gather Current State
+   1. Data and agent configuration is gathered from various sources/APIs in parallel
+2. Create Plan
+   1. Given the current state across the various services, a plan is created based on agent configurations, current Buildkite job queue sizes, and current GCE instances.
+   2. Instances need to be created when there aren't enough online/in-progress agents of a particular configuration to satisfy the needs of its matching queue.
+   3. Agents need to be stopped when the agents have been online for too long (based on their configuration) or when their configuration is out-of-date. This is a soft stop, they will terminate after finishing their current job.
+   4. Instances need to be deleted if they have been stopped (which happens when their agent stops), or when they have been online past their hard stop time (based on configuration).
+3. Execute Plan
+   1. The different types of actions in the plan are executed in parallel. Instance creating and deleting is done in batches to handle spikes quickly.
+
+An error at any step, e.g. when checking current state of GCP instances, will cause the rest of the run to abort.
+
+Because the service gathers data about the total current state and creates a plan based on that state each run, it's reasonably resistant to errors and it's self-healing.
+
+##### Protection against creating too many instances
+
+Creating too many instances in GCP could be costly, so it is worth mentioning here. Since the agent manager itself is stateless, and only looks at the current, external state when determining an execution plan, there is the possibility of creating too many instances.
+
+There are two primary mechanisms to protect against this:
+
+One is usage of GCP quotas. Maintaining reasonable GCP quotas will ensure that we don't create too many instances in a situation where something goes catastrophically wrong during operation. It's an extra failsafe.
+
+The other is built into the agent manager. The agent manager checks both the number of connected agents in Buildkite for a given configuration, as well as the number of instances currently running and being created in GCP. It uses whichever number is greater as the current number of instances.
+
+This is a simple failsafe, but means that a large number of unnecessary instances should only be able to be created in a pretty specific scenario (keep in mind that errors will abort the current agent manager run):
+
+- The GCP APIs (both read and create) are returning success codes
+- The GCP API for listing instances is returning partial/missing/erroneous data, with a success code
+- GCP instances are successfully being created
+- Created GCP instances are unable to connect to Buildkite, or Buildkite Agents API is returning partial/missing/erroneous data
```

Comment:
Buildkite offers [retry logic](https://buildkite.com/docs/pipelines/command-step#retry-attributes) which can be based on the exit code. There are specific exit codes agent connectivity issues.

Reply:
If the agent crashes or stops for any reason, the GCP instance shuts itself down automatically. A new agent will automatically be created, if one is still needed based on the current needs of CI.

I also pull the agent information from Buildkite, and am able to associate running agents with GCP instances. If a GCP instance is online after several minutes but isn't associated with a running agent in Buildkite, I can terminate it so that it will be replaced. This isn't implemented yet but is on the TODO list: #95711